kuri_camera/src/h264_decoder.cpp

```diff
@@ -0,0 +1,348 @@
+  if(picture) {
+    av_free(picture);
+    picture = NULL;
+  }
+  picture_lock.lock();
+
+  if(packet) {
+    av_free_packet(packet);
+    packet = NULL;
+  }
+
+  if (is_alive) {
+    is_alive = false;
+    read_buffer_thread.join();
+    read_frame_thread.join();
+    convert_to_mat_thread.join();
+  }
+
+  got_picture = 0;
+  frame = 0;
+}
+
+// https://timvanoosterhout.wordpress.com/2015/07/02/converting-an-ffmpeg-avframe-to-and-opencv-mat/
+void H264Decoder::avframeToMat(const AVFrame * frame, cv::Mat& image)
+{
+    int width = frame->width;
+    int height = frame->height;
+
+    // Allocate the opencv mat and store its stride in a 1-element array
+    if (image.rows != height || image.cols != width || image.type() != CV_8UC3) {
+      image = cv::Mat(height, width, CV_8UC3);
+    }
+    int cv_linesizes[1];
+    cv_linesizes[0] = image.step1();
+
+    // Convert the color format and write directly to the opencv matrix
+    SwsContext* conversion = sws_getContext(width, height, /*(AVPixelFormat) frame->format*/AV_PIX_FMT_YUV420P, width, height, AV_PIX_FMT_BGR24 /* PIX_FMT_BGR24 */, SWS_FAST_BILINEAR, NULL, NULL, NULL);
+    sws_scale(conversion, frame->data, frame->linesize, 0, height, &image.data, cv_linesizes); // copies data
+    sws_freeContext(conversion);
+}
+
+void H264Decoder::convertToMat() {
+  while (is_alive) {
+    std::unique_lock<std::mutex> picture_lock(picture_mutex);
+    got_picture_cvar.wait(picture_lock);
+    std::unique_lock<std::mutex> cv_image_lock(cv_image_mutex);
+    if (got_picture) {
+      avframeToMat(picture, cv_image);
+      cv_image_recv_timestamp = picture_recv_timestamp;
+      cv_image_lock.unlock();
+      decoded_new_frame.notify_all();
+    }
+    picture_lock.unlock();
+  }
+}
+
+int H264Decoder::getMostRecentFrame(cv::Mat &image, uint64_t &recv_timestamp) {
+  std::unique_lock<std::mutex> cv_image_lock(cv_image_mutex);
+  image = cv_image.clone();
+  recv_timestamp = cv_image_recv_timestamp;
+  cv_image_lock.unlock();
+}
+
+bool H264Decoder::load(std::string hostname, std::string port) {
+
+  codec = avcodec_find_decoder(AV_CODEC_ID_H264);
+  if(!codec) {
+    printf("Error: cannot find the h264 codec: %s:%s\n", hostname.c_str(), port.c_str());
+    return false;
+  }
+
+  codec_context = avcodec_alloc_context3(codec);
+
+  if(codec->capabilities & CODEC_CAP_TRUNCATED) {
+    codec_context->flags |= CODEC_FLAG_TRUNCATED;
+  }
+
+  if(avcodec_open2(codec_context, codec, NULL) < 0) {
+    printf("Error: could not open codec.\n");
+    return false;
+  }
+
+  // Get the TCP endpoint
+  boost::asio::ip::tcp::resolver resolver(io_service);
+  boost::asio::ip::tcp::resolver::query query(hostname, port);
+  iter = resolver.resolve(query);
+
+  std::unique_lock<std::mutex> picture_lock(picture_mutex);
+  // NOTE: depending on your version of libav, you may need the first or second
+  // function below to allocate a frame.
+  picture = av_frame_alloc(); // avcodec_alloc_frame(); // 
```

Comment:
This is a pretty minor API change to paper over. If it's possible, doing a preprocessor check on the library version and using the correct one is preferable (so the code'll just work on whatever's in the 14.04 distribution too).

Reply:
Done